config/v1/types_build.go

```diff
@@ -71,6 +71,14 @@ type BuildDefaults struct {
 	// Resources defines resource requirements to execute the build.
 	// +optional
 	Resources corev1.ResourceRequirements `json:"resources"`
+
+	// ForcePull specifies whether to always check for a newer version of the image
+	// in the registry (true) or allow the use of the local image (if it exists)
+	// even if it is not the latest version (false).
+	// Enabling forcePull is only relevant for tag references since a SHA reference
+	// can by definition only refer to a specific version of an image.
+	// +optional
+	ForcePull *bool `json:"forcePull,omitempty"`
```

Comment:
why is this a pointer? It doesn't make sense as omitempty without true default. If you want that default, add it via annotation.

Reply:
This creates an interesting conundrum (one that I did not anticipate).

Our current `BuildConfig` API uses `bool` primitives to set `ForcePull` on each build strategy. We can't tell if a user left `forcePull` empty or if they explicitly set `forcePull: false`. I don't think we can switch these fields to be a pointer without completely breaking things (ocm, openshift-apiserver, builder image, oc command line, etc.)

In OCP 3.x `forcePull` was introduced as a `BuildOverride`. This lets us force all builds to pull. However, if this was set to `false` then builds fell back to the behavior specified in the `BuildConfig` [1]. We can't do this for `BuildDefault` - the contract for that object is "use these if not specified in the BuildConfig, otherwise preserve what is there". Therefore I think we may be stuck and can't introduce `ForcePull` as a `BuildDefault`.

@bparees @sttts do you agree?

[1] https://github.com/openshift/openshift-controller-manager/blob/master/pkg/build/controller/build/overrides/overrides.go#L35-L52

Reply:
@adambkaplan yeah if we have no way to interpret/distinguish user intent in the individual buildconfig definition, then we can't have a global default.

and @sttts that inability to distinguish user intent coming back to haunt us is why in general i prefer to make booleans be pointers, so that someday in the future if we needed to know whether the admin explicitly chose a value for this field or not, we would be able to do so(suppose for example we later introduce a new field that's supposed to be used only if this field is not set explicitly by the admin).  I know there's disagreement on this philosophy.